### PR TITLE
MLA Alertmanager Health Status

### LINKED ISSUE #8096: MLA Alertmanager Health Status
Currently, we have an `Alertmanager` object for every user cluster to configure the alertmanager. However, this `Alertmanager` is lacking useful status information of if the configuration has been applied successfully or not.
This user story contains the following parts:
1. Have some fields in `Alertmanager.Status` to provide some information about the state of alertmanager.
2. Propegate such information to `Cluster.Status` so that we have the health status of alertmanager for the user cluster.

### PR DESCRIPTION
**Pull request by hdurand0710**
Signed-off-by: Helene Durand <helene@kubermatic.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8096

**Special notes for your reviewer**:
This PR modifies two CRs: `Cluster` and `Alertmanager` to add information about the status of the alert manager configuration update.
The following fields are added:
- Cluster CR:
   - .`extendedHealth.alertmanagerConfig` (HealthStatus, field not present if MLA is disabled on the cluster, same values as other `.extendedHealth` fields)
- AlertmanagerCR:
```
"status": {
        "configStatus": {
            "lastUpdated": <the last time the configuration was successfully applied>,
            "errorMessage": <an error message detailing why it failed if last reconcile was a failure>,
            "status": "True" if last reconciliation was a success, "False" otherwise
        }
    }
```
`ErrorMessage` is reset if the next reconciliation is succesful.

**Examples for Alertmanager CR content:**
   - configuration successfully  applied:
```
"status": {
        "configStatus": {
            "lastUpdated": "2021-10-29T06:33:30Z",
            "status": "True"
        }
    }
```
   - configuration failed to apply:
```
"status": {
        "configStatus": {
            "lastUpdated": "2021-10-29T06:33:30Z",
            "errorMessage": "Get \"http://cortex-alertmanager.mla.svc.cluster.local:8080/api/v1/alerts\": dial tcp: lookup cortex-alertmanager.mla.svc.cluster.local on 8.8.8.8:53: no such host",
            "status": "False"
        }
    }
```

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Add MLA Alertmanager health status in the API /health endpoint (whether the alert manager configuration was successfully applied or not):
- . extendedHealth. alertmanagerConfig
```

### CODE REVIEW

File: pkg/crd/kubermatic/v1/alertmanager.go [hunk 1]
```diff
@@ -51,5 +52,23 @@ type AlertmanagerList struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata,omitempty"`
 
-	Items []Alertmanager `json:"items"`
+	Items  []Alertmanager     `json:"items"`
+	Status AlertManagerStatus `json:"status,omitempty"`
```
Comment (kron4eg) [inline pkg/crd/kubermatic/v1/alertmanager.go]: This is `AlertmanagerList`, I suppose it shouldn't have `Status`.

File: pkg/controller/seed-controller-manager/mla/alertmanager_controller.go [hunk 2]
```diff
@@ -439,3 +479,42 @@ func (r *alertmanagerController) getCurrentAlertmanagerConfig(alertmanagerURL st
...
+	}
+	oldAlertmanager := alertmanager.DeepCopy()
+	oldCluster := cluster.DeepCopy()
+
+	if configErr != nil {
+		alertmanager.Status.ConfigStatus.ErrorMessage = configErr.Error()
+		alertmanager.Status.ConfigStatus.Status = corev1.ConditionFalse
+		status := kubermaticv1.HealthStatusDown
+		cluster.Status.ExtendedHealth.AlertmanagerConfig = &status
+	} else {
```
Comment (zreigz) [inline pkg/controller/seed-controller-manager/mla/alertmanager_controller.go]: you can remove `else` and move the whole  block up as a default 
```
alertmanager.Status.ConfigStatus.ErrorMessage = "" // reset error message
alertmanager.Status.ConfigStatus.Status = corev1.ConditionTrue
alertmanager.Status.ConfigStatus.LastUpdated = metav1.Now()
status := kubermaticv1.HealthStatusUp
cluster.Status.ExtendedHealth.AlertmanagerConfig = &status
if configErr != nil {
		alertmanager.Status.ConfigStatus.ErrorMessage = configErr.Error()
		alertmanager.Status.ConfigStatus.Status = corev1.ConditionFalse
		status := kubermaticv1.HealthStatusDown
		
}

```
```

File: pkg/crd/kubermatic/v1/alertmanager.go (hunk 1 above)
Comment (hdurand0710) [reply pkg/crd/kubermatic/v1/alertmanager.go]: Thanks! I do not even remember how/when I could have added this `Status` in the List. Fixed

File: pkg/controller/seed-controller-manager/mla/alertmanager_controller.go (hunk 2 above)
Comment (hdurand0710) [reply pkg/controller/seed-controller-manager/mla/alertmanager_controller.go]: Done.

File: pkg/controller/seed-controller-manager/mla/alertmanager_controller_test.go [hunk 3]
```diff
@@ -70,16 +72,54 @@ func newTestAlertmanagerReconciler(objects []ctrlruntimeclient.Object, handler h
 	return &reconciler, ts
 }
 
+type alertmanagerConfigStatus struct {
+	clusterStatus      kubermaticv1.HealthStatus                    // Alertmanager config status in the Cluster CR
+	alertmanagerStatus kubermaticv1.AlertManagerConfigurationStatus // Alertmanager config status in the Alertmanager CR
+}
+
+// getAlertmanagerConfigStatusDown returns the needed information when the alertmanager config status is OK:
```
Comment (jiachengxu) [inline pkg/controller/seed-controller-manager/mla/alertmanager_controller_test.go]:
```suggestion
// getAlertmanagerConfigStatusUp returns the needed information when the alertmanager config status is OK:
```

File: pkg/crd/kubermatic/v1/alertmanager.go [hunk 4]
```diff
@@ -53,3 +54,20 @@ type AlertmanagerList struct {
...
 	Items []Alertmanager `json:"items"`
 }
+
+// AlertManagerStatus stores status information about the AlertManager
+type AlertManagerStatus struct {
+	ConfigStatus AlertManagerConfigurationStatus `json:"configStatus,omitempty"`
+}
+
+// AlertManagerConfigurationStatus stores status information about the AlertManager configuration
+
```
Comment (jiachengxu) [inline pkg/crd/kubermatic/v1/alertmanager.go]: Please remove this empty line

**Pull request merged**